compiler/rustc_infer/src/infer/error_reporting/suggest.rs

```diff
@@ -351,6 +351,59 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         }
     }
 
+    pub(super) fn suggest_function_pointers(
+        &self,
+        cause: &ObligationCause<'tcx>,
+        span: Span,
+        exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
+        diag: &mut Diagnostic,
+    ) {
+        debug!("suggest_function_pointers(cause={:?}, exp_found={:?})", cause, exp_found);
+        let ty::error::ExpectedFound { expected, found } = exp_found;
+        let expected_inner = expected.peel_refs();
+        let found_inner = found.peel_refs();
+        if !expected_inner.is_fn() || !found_inner.is_fn() {
+            return;
+        }
+        diag.note("different `fn` items have unique types, even if their signatures are the same");
+        match (&expected_inner.kind(), &found_inner.kind()) {
+            (ty::FnPtr(sig), ty::FnDef(did, substs)) => {
+                let sig1 = &(self.normalize_fn_sig)(*sig);
+                let sig2 = &(self.normalize_fn_sig)(found_inner.fn_sig(self.tcx));
+                if sig1 != sig2 {
+                    return;
+                }
+                let fn_name = self.tcx.def_path_str_with_substs(*did, substs);
+                let (msg, sugg) = match (expected.is_ref(), found.is_ref()) {
```

Comment:
we should check [`is_suggestable`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/trait.IsSuggestable.html#tymethod.is_suggestable) on the `sig`. the `infer_suggestable` arg can be true.